src/language-markdown/printer-markdown.js

```diff
@@ -52,6 +56,43 @@ const SIBLING_NODE_TYPES = new Set([
   "definition",
   "footnoteDefinition",
 ]);
+// https://en.wikipedia.org/wiki/Line_breaking_rules_in_East_Asian_languages
+/**
+ * The set of characters that must not immediately precede a line break
+ *
+ * e.g. `"（"`
+ *
+ * - Bad:  `"檜原村（\nひのはらむら）"`
+ * - Good: `"檜原村\n（ひのはらむら）"` or ``"檜原村（ひ\nのはらむら）"`
+ */
+const noBreakAfterCharacterSet = new Set(
+  "$(£¥·'\"〈《「『【〔〖〝﹙﹛＄（［｛￡￥[{‵︴︵︷︹︻︽︿﹁﹃﹏〘｟«"
+);
+/**
+ * The set of characters that must not immediately follow a line break
+ *
+ * e.g. `"）"`
+ *
+ * - Bad:  `"檜原村（ひのはらむら\n）以外には、"`
+ * - Good: `"檜原村（ひのはらむ\nら）以外には、"` or `"檜原村（ひのはらむら）\n以外には、"`
+ */
+const noBreakBeforeCharacterSet = new Set(
+  "!%),.:;?]}¢°·'\"†‡›℃∶、。〃〆〕〗〞﹚﹜！＂％＇），．：；？］｝～–—•〉》」︰︱︲︳﹐﹑﹒﹓﹔﹕﹖﹘︶︸︺︼︾﹀﹂﹗｜､』】〙〟｠»ヽヾーァィゥェォッャュョヮヵヶぁぃぅぇぉっゃゅょゎゕゖㇰㇱㇲㇳㇴㇵㇶㇷㇸㇹㇺㇻㇼㇽㇾㇿ々〻‐゠〜～‼⁇⁈⁉・"
+);
+/**
+ * The set of characters whose surrounding newline may be converted to Space
+ *
+ * - ASCII punctuation marks
+ */
+const lineBreakBetweenTheseAndCJKConvertToSpaceSymbolSet = new Set(
+  "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"
+);
+
+/**
+ * Unicode punctuation character defined in [CommonMark Spec](https://spec.commonmark.org/0.30/#unicode-punctuation-character)
+ */
+const punctuationRegex =
+  /(?:\p{Pc}|\p{Pd}|\p{Pe}|\p{Pf}|\p{Pi}|\p{Po}|\p{Ps})/u;
```

Comment:
It's the same regex that already exists in `utils.js`, right? Let's just export if from there.

Reply:
Sorry could you tell me where? I could not find one.

Reply:
It's called the same there: punctuationRegex

Reply:
@thorn0 First time I knew about it.I will divert it.